pkg/controller/groupbinding/groupbinding_controller.go

```diff
@@ -0,0 +1,324 @@
+		c.workqueue.Forget(obj)
+		klog.Infof("Successfully synced %s:%s", "key", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the Foo resource
+// with the current status of the resource.
+func (c *Controller) reconcile(key string) error {
+
+	groupBinding, err := c.groupBindingLister.Get(key)
+	if err != nil {
+		// The user may no longer exist, in which case we stop
+		// processing.
+		if errors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("groupbinding '%s' in work queue no longer exists", key))
+			return nil
+		}
+		klog.Error(err)
+		return err
+	}
+	if groupBinding.ObjectMeta.DeletionTimestamp.IsZero() {
+
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object.
+		if !sliceutil.HasString(groupBinding.Finalizers, finalizer) {
+			groupBinding.ObjectMeta.Finalizers = append(groupBinding.ObjectMeta.Finalizers, finalizer)
+			if groupBinding, err = c.ksClient.IamV1alpha2().GroupBindings().Update(groupBinding); err != nil {
+				return err
+			}
+			// Skip reconcile when groupbinding is updated.
+			return nil
+		}
+	} else {
+		// The object is being deleted
+		if sliceutil.HasString(groupBinding.ObjectMeta.Finalizers, finalizer) {
+			if err = c.bindUser(groupBinding); err != nil {
+				klog.Error(err)
+				return err
+			}
+
+			// remove our finalizer from the list and update it.
+			groupBinding.Finalizers = sliceutil.RemoveString(groupBinding.ObjectMeta.Finalizers, func(item string) bool {
+				return item == finalizer
+			})
+
+			if groupBinding, err = c.ksClient.IamV1alpha2().GroupBindings().Update(groupBinding); err != nil {
+				return err
+			}
+		}
+		// Our finalizer has finished, so the reconciler can do nothing.
+		return nil
+	}
+
+	if err = c.bindUser(groupBinding); err != nil {
+		klog.Error(err)
+		return err
+	}
+
+	c.recorder.Event(groupBinding, corev1.EventTypeNormal, successSynced, messageResourceSynced)
+	return nil
+}
+
+func (c *Controller) Start(stopCh <-chan struct{}) error {
+	return c.Run(4, stopCh)
+}
+
+// Udpate user's Group property. So no need to query user's groups when authorizing.
+func (c *Controller) bindUser(groupBinding *iamv1alpha2.GroupBinding) error {
+
+	users := make([]string, 0)
+	// Ignore the user if the user if being deleted.
+	for _, u := range groupBinding.Users {
+		if user, err := c.ksClient.IamV1alpha2().Users().Get(u, metav1.GetOptions{}); err == nil && user.ObjectMeta.DeletionTimestamp.IsZero() {
+			users = append(users, u)
+		}
+	}
+
+	// Nothing to do
+	if len(users) == 0 {
+		return nil
+	}
+
+	// Get all GroupBindings and check whether user exists in the Group.
+	listOptions := metav1.ListOptions{}
+	groupBindingList, err := c.ksClient.IamV1alpha2().GroupBindings().List(listOptions)
```

Comment:
Check all group binding is unnecessary. Only the current changes need to be applied.

Reply:
Refined. Please check again.